File: tests/mirror_run.js
/* Mirror-Runner: <robot> <seed> <steps> <genomeSeed>
 * Fährt die JS-Env mit festem Genom und druckt Fit + Obs als JSON
 * (volle double-Genauigkeit) — Vergleichspartner für kaggle/mirror_run.py. */
'use strict';
const fs = require('fs');
const path = require('path');
const vm = require('vm');
const SRC = path.join(__dirname, '..', 'sim', 'src');
for (const f of ['rng.js', 'nn.js', 'envs.js']) {
  vm.runInThisContext(fs.readFileSync(path.join(SRC, f), 'utf8'), { filename: f });
}
const TF = globalThis.TF07;
const [robot, seed, steps, genomeSeed] = [process.argv[2], +process.argv[3], +process.argv[4], +process.argv[5]];
const env = TF.makeEnv(robot, seed);
const nAct = TF.nn.ARCHS[robot][3];
const genome = TF.nn.initGenome(genomeSeed, TF.nn.ARCHS[robot][0], nAct);
let done = false;
for (let i = 0; i < steps && !done; i++) {
  const a = TF.nn.forward(genome, env.getObs(), nAct);
  if (robot === 'duck') done = env.step(a[0], a[1]).done;
  else if (robot === 'arm') done = env.step(a[0], a[1], a[2], a[3], a[4]).done;
  else done = env.step(a).done;
}
const obs = env.getObs();
console.log(JSON.stringify({
  fit: env.fit,
  steps: env.steps,
  reached: env.reached || env.delivered || 0,
  obs: Array.from(obs)
}));

File: kaggle/mirror_run.py
#!/usr/bin/env python3
"""Mirror-Runner (Python): <robot> <seed> <steps> <genomeSeed>
Exakt derselbe Lauf wie tests/mirror_run.js — für den Determinismus-Vergleich."""
import json
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np
from robofield_common import ARCHS, init_genome, forward_single
from robofield_envs import SCALAR

robot, seed, steps, genome_seed = sys.argv[1], int(sys.argv[2]), int(sys.argv[3]), int(sys.argv[4])
env = SCALAR[robot](seed)
n_act = ARCHS[robot][3]
genome = init_genome(genome_seed, ARCHS[robot][0], n_act)
done = False
for _ in range(steps):
    if done:
        break
    a = forward_single(genome, np.array(env.get_obs()), n_act)
    if robot == "duck":
        _, done = env.step(a[0], a[1])
    elif robot == "arm":
        _, done = env.step(a[0], a[1], a[2], a[3], a[4])
    else:
        _, done = env.step(a)
out = {
    "fit": env.fit,
    "steps": env.steps,
    "reached": getattr(env, "reached", getattr(env, "delivered", 0)),
    "obs": [float(x) for x in env.get_obs()],
}
print(json.dumps(out))
